sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/FeatureManager.java

```diff
@@ -73,45 +91,178 @@ public Mono<Boolean> isEnabledAsync(String feature) {
+
+        if (!StringUtils.hasText(featureName)) {
+            throw new IllegalArgumentException("Feature Variant name can not be empty or null.");
+        }
+
+        Feature feature = featureManagementConfigurations.getFeatureManagement().get(featureName);
+
+        if (feature == null) {
+            throw new FeatureManagementException("The Feature " + featureName + " can not be found.");
+        }
+
+        validateVariant(feature, featureName);
+
+        // Disabled?
+        if (!feature.getEvaluate() && StringUtils.hasText(feature.getAllocation().getDefautlWhenDisabled())) {
+            return variantAssignment.getVariant(feature.getVariants().values(),
+                feature.getAllocation().getDefautlWhenDisabled()).single();
+        } else if (!feature.getEvaluate()) {
+            return Mono.justOrEmpty(null);
+        } else if (feature.getEnabledFor().size() == 0) {
+            return variantAssignment.assignVariant(feature.getAllocation(), feature.getVariants().values());
+        }
+
+        List<Mono<Boolean>> results = new ArrayList<>();
+
+        for (FeatureFilterEvaluationContext featureFilter : feature.getEnabledFor().values()) {
+            if (StringUtils.hasText(featureFilter.getName())) {
+                results.add(isFeatureOn(featureFilter, feature.getKey(), featureContext));
+            }
+        }
+
+        Mono<Boolean> isEnabled;
+
+        // All Filters must be true
+        if (ALL_REQUIREMENT_TYPE.equals(feature.getRequirementType())) {
+            isEnabled = Flux.merge(results).reduce((a, b) -> a && b).single();
+        } else {
+            // Any Filter must be true
+            isEnabled = Flux.merge(results).reduce((a, b) -> a || b).single();
+        }
+
+        return isEnabled.flatMap(enabled -> {
+            if (!enabled && StringUtils.hasText(feature.getAllocation().getDefautlWhenDisabled())) {
+                return variantAssignment.getVariant(feature.getVariants().values(),
+                    feature.getAllocation().getDefautlWhenDisabled()).single();
+            } else if (!enabled) {
+                return Mono.justOrEmpty(null);
+            }
+
+            return variantAssignment.assignVariant(feature.getAllocation(), feature.getVariants().values());
+        });
+
+    }
+
+    private void validateVariant(Feature feature, String featureName) {
+        if (feature.getVariants() == null || feature.getVariants().size() == 0) {
+            throw new FeatureManagementException("The " + feature.getKey() + " has no assigned Variants.");
+        }
+
+        for (VariantReference variant : feature.getVariants().values()) {
+            if (!StringUtils.hasText(variant.getName())) {
+                throw new FeatureManagementException("Variant needs a name");
+            }
+
+            if (variant.getConfigurationValue() == null && variant.getConfigurationReference() == null) {
```

Comment:
.NET allows both to be null, and the actual `Variant` returned will just have a null value.

Reply:
In most languages returning null is as an expected result is frowned upon. Is there a good reason we are doing it. I would rather not. I would prefer if we don't even in the few places that I currently do it.

Reply:
I don't remember the whole conversation around it, but the main reason was that variants are a new concept, so we want to make it possible for people to start off with very simple allocation/variants sections without needing to understand all of the concepts to use it. People may even just want to create multiple variants and just make decisions in their code based only on their names. I think it's just important that it's properly documented, and we add logs to clarify what's going on. I'm not fully aware of the conventions around returning null in c# vs java though.